File: src/frontend/src/api/terms.ts
import { api } from "./client";
import type { ProductCategory } from "./products";

export interface TermResponse {
  id: number;
  circle_group_id: number;
  /** ISO datetime — class date *and* wall-clock start time (`2026-03-12T17:30:00`). */
  occurs_on: string;
  description: string | null;
  created_at: string;
  updated_at: string;
}

export interface CreateTermRequest {
  circle_group_id: number;
  occurs_on: string;
  description?: string;
}

export interface NeededItemResponse {
  id: number;
  term_id: number;
  product_id: number;
  product_name: string;
  product_category: ProductCategory;
  description: string | null;
  created_at: string;
  updated_at: string;
}

export interface CreateNeededItemRequest {
  term_id: number;
  product_id: number;
  description?: string;
}

export function getTerms(circleGroupId: number): Promise<TermResponse[]> {
  return api.get(`/terms?circle_group_id=${circleGroupId}`);
}

export function getTerm(id: number): Promise<TermResponse> {
  return api.get(`/terms/${id}`);
}

export function createTerm(request: CreateTermRequest): Promise<TermResponse> {
  return api.post("/terms", request);
}

export interface UpdateTermRequest {
  occurs_on?: string;
  description?: string;
}

export function updateTerm(id: number, request: UpdateTermRequest): Promise<TermResponse> {
  return api.patch(`/terms/${id}`, request);
}

export function getNeededItems(termId: number): Promise<NeededItemResponse[]> {
  return api.get(`/needed-items?term_id=${termId}`);
}

export function getNeededItem(id: number): Promise<NeededItemResponse> {
  return api.get(`/needed-items/${id}`);
}

export function createNeededItem(request: CreateNeededItemRequest): Promise<NeededItemResponse> {
  return api.post("/needed-items", request);
}

export interface UpdateNeededItemRequest {
  product_id?: number;
  description?: string;
}

export function updateNeededItem(
  id: number,
  request: UpdateNeededItemRequest,
): Promise<NeededItemResponse> {
  return api.patch(`/needed-items/${id}`, request);
}

export function deleteNeededItem(id: number): Promise<void> {
  return api.delete(`/needed-items/${id}`);
}

File: src/backend/app/groups/router/terms.py
"""`/api/terms` and `/api/needed-items` routes."""

from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Depends, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.auth_deps import Principal, require_any
from app.db import get_db
from app.groups import service
from app.groups.schemas import (
    CreateNeededItemRequest,
    CreateTermRequest,
    NeededItemResponse,
    TermResponse,
    UpdateNeededItemRequest,
    UpdateTermRequest,
)
from app.users.service import get_profile_by_principal

router = APIRouter(tags=["groups"])

DbSession = Annotated[AsyncSession, Depends(get_db)]
ReadPrincipal = Annotated[Principal, Depends(require_any("READ", "mcp:read"))]
EditPrincipal = Annotated[Principal, Depends(require_any("EDIT", "mcp:edit"))]


# --- Term / NeededItem ----------------------------------------------------------


@router.post("/api/terms", response_model=TermResponse, status_code=status.HTTP_201_CREATED)
async def create_term(
    body: CreateTermRequest, db: DbSession, principal: EditPrincipal
) -> TermResponse:
    term = await service.create_term(db, principal, body)
    return TermResponse.model_validate(term)


@router.get("/api/terms", response_model=list[TermResponse])
async def list_terms(
    circle_group_id: int, db: DbSession, principal: ReadPrincipal
) -> list[TermResponse]:
    terms = await service.list_terms(db, circle_group_id)
    return [TermResponse.model_validate(term) for term in terms]


@router.get("/api/terms/{term_id}", response_model=TermResponse)
async def get_term(term_id: int, db: DbSession, principal: ReadPrincipal) -> TermResponse:
    term = await service.get_term(db, term_id)
    return TermResponse.model_validate(term)


@router.patch("/api/terms/{term_id}", response_model=TermResponse)
async def update_term(
    term_id: int, body: UpdateTermRequest, db: DbSession, principal: EditPrincipal
) -> TermResponse:
    profile = await get_profile_by_principal(db, principal)
    term = await service.update_term(db, term_id, profile.party_id, body)
    return TermResponse.model_validate(term)


@router.post(
    "/api/needed-items", response_model=NeededItemResponse, status_code=status.HTTP_201_CREATED
)
async def create_needed_item(
    body: CreateNeededItemRequest, db: DbSession, principal: EditPrincipal
) -> NeededItemResponse:
    needed_item = await service.create_needed_item(db, principal, body)
    return NeededItemResponse.model_validate(needed_item)


@router.get("/api/needed-items", response_model=list[NeededItemResponse])
async def list_needed_items(
    term_id: int, db: DbSession, principal: ReadPrincipal
) -> list[NeededItemResponse]:
    needed_items = await service.list_needed_items(db, term_id)
    return [NeededItemResponse.model_validate(needed_item) for needed_item in needed_items]


@router.get("/api/needed-items/{needed_item_id}", response_model=NeededItemResponse)
async def get_needed_item(
    needed_item_id: int, db: DbSession, principal: ReadPrincipal
) -> NeededItemResponse:
    needed_item = await service.get_needed_item(db, needed_item_id)
    return NeededItemResponse.model_validate(needed_item)


@router.patch("/api/needed-items/{needed_item_id}", response_model=NeededItemResponse)
async def update_needed_item(
    needed_item_id: int,
    body: UpdateNeededItemRequest,
    db: DbSession,
    principal: EditPrincipal,
) -> NeededItemResponse:
    profile = await get_profile_by_principal(db, principal)
    needed_item = await service.update_needed_item(db, needed_item_id, profile.party_id, body)
    return NeededItemResponse.model_validate(needed_item)


@router.delete("/api/needed-items/{needed_item_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_needed_item(needed_item_id: int, db: DbSession, principal: EditPrincipal) -> None:
    profile = await get_profile_by_principal(db, principal)
    await service.soft_delete_needed_item(db, needed_item_id, profile.party_id)
    return None
